### DISCUSSION TITLE
Creating an Organisation: What are Key, Keys.EncryptedPrivateKey and Keys.PublicKey

### DISCUSSION THREAD

**Discussion by uedvt359** [dani-garcia/vaultwarden#3648] [Q&A]
I'm trying to automate creation of organisations through the API. The `/api/organizations` endpoint takes the following parameters:

- BillingEmail: obvious
- CollectionName: encrypted name of the default collection?
- Key: String,
- Name: obvious
- Keys: {
  - EncryptedPrivateKey: String,
  - PublicKey: String,
PlanType: ignored

I've looked at both [vaultwarden's source](), and a [python client implementation](https://github.com/corpusops/bitwardentools/blob/5aa88303b3d7a7c469cd333390158cee642f2646/src/bitwardentools/client.py#L1336), but I can't figure out which keys are which. 

- I *think* Key is the (freshly generated) AES key, which is used to en/decrypt the default collection, but encrypted with the creating user's RSA PrivateKey?
- Keys.EncryptedPrivateKey is something AES-encrypted? I tried following Organization::new(), but it just set them as attributes
- It also looks to me that Keys is optional - what happens when these are not supplied? What's the difference if I supply them?

**Comment by BlackDex** [comment]
Vaultwarden does nothing with the encryption it self. It is all handled on the client side.
See https://bitwarden.com/help/what-encryption-is-used/ for information in regards to what encryption is used and how.

**Reply by uedvt359** [reply]
Thanks, that explains what I found in vaultwarden's code. Do you happen to know what key material is stored in Keys.EncryptedPrivateKey/Keys.PublicKey? A freshly generated 2kbit RSA pair? Which key is used to encrypt it?

**Reply by BlackDex** [reply]
I suggest to read the security paper available on the link i posted. That will explain it much better then i can (if i can at all).

**Comment by uedvt359** [comment]
I managed to piece it together, thanks blackdex for the pointers. this below is the meat of it, probably missing some global vars/imports

```
from Crypto.PublicKey import RSA # PyPI pycryptodome
from Crypto.Cipher import PKCS1_OAEP, AES
from Crypto.Util.Padding import pad
def create_org():
        org_name = "my cool org"
        default_collection = f"{org_name} default collection"
        superadmin = "owner@company.com"

        shareKey = secrets.token_bytes(64) # used as 256 bit AES key
        enc_shareKey = rsa_encrypt(shareKey, base64.b64decode(superadmin_pubkey))

        asym_key = RSA.generate(2048)
        pk = asym_key.public_key().export_key("DER")
        sk = asym_key.export_key("DER", pkcs=8)
        enc_sk = aes_encrypt(sk, shareKey)

        collection = aes_encrypt(default_collection.encode(), shareKey)

        r = session.post("/organizations", json={
            "billingEmail": superadmin,
            "collectionName": collection,
            "key": enc_shareKey,
            "name": org_name,
            "keys": {
                "publicKey": base64.b64encode(pk).decode(),
                "encryptedPrivateKey": enc_sk,
            },
            "planType": 0, # ignored
        })

def rsa_encrypt(plaintext, key):
    cipher = PKCS1_OAEP.new(RSA.import_key(key)).encrypt(plaintext) # expects DER
    b64_ct = base64.b64encode(cipher).decode()
    return f"4.{b64_ct}" # 4 is bitwarden's magic constant for RSA encryption

def aes_encrypt(plaintext, key):
    enc_key, mac_key = key[:32], key[32:]

    aes = AES.new(enc_key, AES.MODE_CBC)

    ct = aes.encrypt(pad(plaintext, AES.block_size))
    mac = hmac.new(mac_key, aes.iv+ct, hashlib.sha256)

    iv = base64.b64encode(aes.iv).decode()
    ct = base64.b64encode(ct).decode()
    mac = base64.b64encode(mac.digest()).decode()
    return f"2.{iv}|{ct}|{mac}" # 2 is bitwarden's magic constant for AES encryption
```
